hi-ml/src/health_ml/utils/checkpoint_handler.py

```diff
@@ -69,10 +76,56 @@ def get_checkpoint_to_test(self) -> Path:
+            # Handling low-priority preemption: If we are recovering from a preemption that appeared after training was
+            # finished, there will not be an inference checkpoint available on disk. In that case, we need to download
+            # it from the AzureML run.
+            logging.info("Trying to find an inference checkpoint in AzureML.")
+            downloaded_checkpoint = self.download_inference_checkpoint()
+            if downloaded_checkpoint is not None:
+                logging.info(f"Using a checkpoint found in the AzureML run: {downloaded_checkpoint}")
+                return downloaded_checkpoint
+            raise FileNotFoundError("No inference checkpoint file found locally nor in AzureML.")
         elif self.trained_weights_path:
             # Model was not trained, check if there is a local weight path.
             logging.info(f"Using pre-trained weights from {self.trained_weights_path}")
             return self.trained_weights_path
         raise ValueError("Unable to determine which checkpoint should be used for testing.")
+
+    def get_relative_inference_checkpoint_path(self) -> Path:
+        """Returns the path of the model's inference checkpoint, relative to the container's output folder."""
+        expected_checkpoint_path = self.container.get_checkpoint_to_test()
+        try:
+            # Find the Unix-style path of the checkpoint relative to the outputs folder, this will be the path
+            # where we find the file in AzureML.
+            return expected_checkpoint_path.relative_to(self.container.outputs_folder)
+        except ValueError:
+            raise ValueError("Inference checkpoint path should be relative to the container's output folder. "
+                             f"Checkpoint path: {expected_checkpoint_path}, "
+                             f"output folder: {self.container.outputs_folder}")
+
+    def download_inference_checkpoint(self, download_folder: Optional[Path] = None) -> Optional[Path]:
+        """
+        For low-priority preemption that occured after training, try to download the inference checkpoint if that
+        is available in the AzureML run from a previous incarnation of the job. Downloads go into the `download_folder`.
+
+        :param download_folder: The folder where the checkpoints should be downloaded to. If not provided, use a
+            temp folder.
+        :return: The path to a downloaded inference checkpoint.
+        """
+        # This logic will only trigger in AzureML. Download should only happen once per node.
+        if is_running_in_azure_ml() and is_global_rank_zero():
+            download_folder = download_folder or Path(tempfile.mkdtemp())
+            inference_checkpoint_azureml_path = (
+                Path(DEFAULT_AML_UPLOAD_DIR) / self.get_relative_inference_checkpoint_path()
+            ).as_posix()
+            highest_epoch_checkpoint = download_highest_epoch_checkpoint(
```

Comment:
I think the naming is justified. In the AzureML run, we can have multiple inference checkpoints, for example there can be `retry_001/best_val.ckpt`, `retry_002/best_val.ckpt`. These are written in different retries as the validation set metrics improve. Among those checkpoints, we choose the one with the highest epoch, as this will be the most recent one.